File: ArrowMove/src/game/controller/CommandLineController.java
package game.controller;

import java.io.*;
import java.lang.ProcessBuilder.Redirect;

import game.data.CommandData;
import game.data.Data;
import game.data.TextCommandData;

//プロセス用のController
public class CommandLineController extends Controller<Data> {
	public static final String READ_END = "end";
	private BufferedReader reader;
	private BufferedWriter writer;
	public CommandLineController() {
		super("commandLineController");
	}

	@Override
	public void open() {
		reader =new BufferedReader(new InputStreamReader(System.in));
		writer =new BufferedWriter(new OutputStreamWriter(System.out));
	}

	@Override
	public TextCommandData controll(Data data) {
		String command = "";
		try {
			//ゲームのデータを出力
			for (String s:data.toString().split("\n")) {
				writer.write(s);
				writer.newLine();
			}
			writer.flush();
			//Processの行動を入力
			String line = null;
			while(true){

				line = reader.readLine();
				if(line==null || isEndLine(line))break;
				command += line + "\n";
			}



		} catch (IOException e) {
			e.printStackTrace();
		}
		return new TextCommandData(this,command);
	}

	private boolean isEndLine(String line)  {
		return line.equals(READ_END);
	}

	@Override
	public void close() {
		/*try {
			reader.close();
			writer.close();
		} catch (IOException e) {
			e.printStackTrace();
		}*/


	}

}

File: ArrowMove/src/game/controller/CommandLineController.kt
package game.controller

import java.io.*
import java.lang.ProcessBuilder.Redirect

import game.data.CommandData
import game.data.Data
import game.data.TextCommandData

//プロセス用のController
class CommandLineController : Controller<Data>("commandLineController") {
    private var reader: BufferedReader? = null
    private var writer: BufferedWriter? = null

    @Override
    override fun open() {
        reader = BufferedReader(InputStreamReader(System.`in`))
        writer = BufferedWriter(OutputStreamWriter(System.out))
    }

    @Override
    override fun controll(data: Data): TextCommandData {
        var command = ""
        try {
            //ゲームのデータを出力
            for (s in data.toString().split("\n")) {
                writer!!.write(s)
                writer!!.newLine()
            }
            writer!!.flush()
            //Processの行動を入力
            var line: String? = null
            while (true) {

                line = reader!!.readLine()
                if (line == null || isEndLine(line!!)) break
                command += line!! + "\n"
            }


        } catch (e: IOException) {
            e.printStackTrace()
        }

        return TextCommandData(this, command)
    }

    private fun isEndLine(line: String): Boolean {
        return line.equals(READ_END)
    }

    @Override
    override fun close() {
        /*try {
			reader.close();
			writer.close();
		} catch (IOException e) {
			e.printStackTrace();
		}*/


    }

    companion object {
        val READ_END = "end"
    }

}
